### PR TITLE
Change offlineasm to emit more efficient LLInt code.

### PR DESCRIPTION
**Pull request by MenloDorian**
#### daa5f7a102774147288388f3a5553326a3b72f45
<pre>
Change offlineasm to emit more efficient LLInt code.
<a href="https://bugs.webkit.org/show_bug.cgi?id=241856">https://bugs.webkit.org/show_bug.cgi?id=241856</a>

Reviewed by Yusuke Suzuki.

1. Ruby treats numeric 0 as truthy.  However, there&apos;s a test in arm64LowerMalformedLoadStoreAddresses
   which assumes a value of 0 would be false.  As a result, we see offlineasm emit inefficient LLInt
   code like this:
    &quot;.loc 3 821\n&quot;        &quot;movz x16, #0 \n&quot;                    // LowLevelInterpreter64.asm:821
                          &quot;add x13, x3, x16 \n&quot;
                          &quot;ldr x0, [x13] \n&quot;

  ...  instead of this:
    &quot;.loc 3 821\n&quot;        &quot;ldr x0, [x3] \n&quot;                    // LowLevelInterpreter64.asm:821

   This patch fixes this.

2. offlineasm&apos;s emitARM64MoveImmediate chooses to use `movn` instead of `movz` based on whether a
   64-bit value is negative or not.  Instead, it should be making that decision based on the number of
   halfwords (16-bits) in the value that is 0xffff vs 0.  As a result, offlineasm emits code like this:
    &quot;.loc 1 1638\n&quot;       &quot;movn x27, #1, lsl #48 \n&quot;           // LowLevelInterpreter.asm:1638
                          &quot;movk x27, #0, lsl #32 \n&quot;
                          &quot;movk x27, #0, lsl #16 \n&quot;
                          &quot;movk x27, #0 \n&quot;

  ...  instead of this:
    &quot;.loc 1 1638\n&quot;       &quot;movz x27, #65534, lsl #48 \n&quot;       // LowLevelInterpreter.asm:1638

   This patch fixes this.

3. offlineasm is trivially assuming the range of immediate offsets for ldr/str instructions is
   [-255..4095].  However, that&apos;s only the range for byte sized load-stores.  For 32-bit, the range
   is actually [-255..16380].  For 64-bit, the range is actually [-255..32760].  As a result,
    offlineasm emits code like this:
    &quot;.loc 1 633\n&quot;        &quot;movn x16, #16383 \n&quot;                // LowLevelInterpreter.asm:633
    &quot;.loc 1 1518\n&quot;       &quot;and x3, x3, x16 \n&quot;                 // LowLevelInterpreter.asm:1518
    &quot;.loc 1 1519\n&quot;       &quot;movz x16, #16088 \n&quot;                // LowLevelInterpreter.asm:1519
                          &quot;add x17, x3, x16 \n&quot;
                          &quot;ldr x3, [x17] \n&quot;

  ...  instead of this:
    &quot;.loc 1 633\n&quot;        &quot;movn x17, #16383 \n&quot;                // LowLevelInterpreter.asm:633
    &quot;.loc 1 1518\n&quot;       &quot;and x3, x3, x17 \n&quot;                 // LowLevelInterpreter.asm:1518
    &quot;.loc 1 1519\n&quot;       &quot;ldr x3, [x3, #16088] \n&quot;            // LowLevelInterpreter.asm:1519

   This patch fixes this for 64-bit and 32-bit load-stores.  16-bit load-stores also has a wider
   range, but for now, it will continue to use the conservative range.

   This patch also introduces an `isMalformedArm64LoadAStoreAddress` so that this range check can be
   done consistently in all the places that checks for it.

4. offlineasm is eagerly emitting no-op arguments in instructions, e.g. &quot;lsl #0&quot;, and adding 0.
   As a result, offlineasm emits code like this:
    &quot;.loc 3 220\n&quot;        &quot;movz x13, #51168, lsl #0 \n&quot;        // LowLevelInterpreter64.asm:220
                          &quot;add x17, x1, x13, lsl #0 \n&quot;
                          &quot;ldr w4, [x17, #0] \n&quot;

  ...  instead of this:
    &quot;.loc 3 220\n&quot;        &quot;movz x13, #51168 \n&quot;                // LowLevelInterpreter64.asm:220
                          &quot;add x17, x1, x13 \n&quot;
                          &quot;ldr w4, [x17] \n&quot;

   This unnecessary arguments are actually very common throughout the emitted LLIntAssembly.h.

   This patch removes these unnecessary arguments, which makes the emitted LLInt code more human
   readable due to less clutter.

This patch has passed the testapi and JSC stress tests with a Release build on an M1 Mac.

I also manually verified that the emitARM64MoveImmediate code is working properly by
hacking up LowLevelInterpreter64.asm to emit moves of constants of different values in
the ranges, and for load-store instructions of different sizes, and visually inspecting
the emitted code.

* Source/JavaScriptCore/offlineasm/arm64.rb:

Canonical link: <a href="https://commits.webkit.org/251771@main">https://commits.webkit.org/251771@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295766">https://svn.webkit.org/repository/webkit/trunk@295766</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/JavaScriptCore/offlineasm/arm64.rb [hunk 1]
```diff
@@ -264,29 +283,42 @@ def arm64Operand(kind)
...
-            if opcode =~ /q$/ and $currentSettings["ADDRESS64"]
-                malformed ||= operand.offset.value % 8
+def isMalformedArm64LoadAStoreAddress(opcode, operand)
+    malformed = false
+    if operand.is_a? Address
+        case opcode
+        when "loadp", "storep", "loadq", "storeq"
+            if $currentSettings["ADDRESS64"]
+                malformed ||= (not (-255..32760).include? operand.offset.value)
+                malformed ||= (operand.offset.value % 8 != 0)
```
Comment (Constellation) [inline Source/JavaScriptCore/offlineasm/arm64.rb]: `(operand.offset.value % 8).nonzero?` would be more like Ruby-ish code.

File: Source/JavaScriptCore/offlineasm/arm64.rb [hunk 2]
```diff
@@ -264,29 +283,42 @@ def arm64Operand(kind)
...
+            else
+                malformed ||= (not (-255..16380).include? operand.offset.value)
             end
+        when "loadd", "stored"
+            malformed ||= (not (-255..32760).include? operand.offset.value)
+            malformed ||= (operand.offset.value % 8 != 0)
```
Comment (Constellation) [inline Source/JavaScriptCore/offlineasm/arm64.rb]: Ditto.

**Review by Constellation — APPROVED** [review]
r=me

File: Source/JavaScriptCore/offlineasm/arm64.rb (hunk 1 above)
Comment (MenloDorian) [reply Source/JavaScriptCore/offlineasm/arm64.rb]: I updated to to (not (operand.offset.value % 8).zero?).  It looks like nonzero? does not return a bool.

**Pull request merged**